src/utils/datatool/datatool.cpp

```diff
@@ -35,6 +38,88 @@ int createNavMesh(const std::string& meshFile, const std::string& navmeshFile) {
   return 0;
 }
 
+int createGibsonSemanticMesh(const std::string& objFile,
+                             const std::string& idsFile,
+                             const std::string& semMeshFile) {
+  LOG(INFO) << "createGibsonSemanticMesh";
+  tinyobj::attrib_t attrib;
+  std::vector<tinyobj::shape_t> shapes;
+  std::vector<tinyobj::material_t> materials;
+
+  std::string warn;
+  std::string err;
+
+  bool ret = tinyobj::LoadObj(&attrib, &shapes, &materials, &warn, &err,
+                              objFile.c_str());
+  if (!warn.empty()) {
+    LOG(WARNING) << warn;
+  }
+  if (!err.empty()) {
+    LOG(ERROR) << err;
+  }
+  if (!ret) {
+    LOG(ERROR) << "Failed to load " << objFile;
+    return 1;
+  }
+
+  std::ifstream file(idsFile, std::ios::binary | std::ios::ate);
+  std::streamsize size = file.tellg();
+  file.seekg(0, std::ios::beg);
+
+  std::vector<unsigned short> objectId(size / sizeof(short));
+  file.read(reinterpret_cast<char*>(objectId.data()), size);
+  if (!file) {
+    LOG(ERROR) << "Failed to load " << idsFile;
+    return 2;
+  }
+
+  size_t numVerts = attrib.vertices.size() / 3;
+
+  std::ofstream f(semMeshFile, std::ios::out | std::ios::binary);
+  f << "ply" << std::endl;
+  f << "format binary_little_endian 1.0" << std::endl;
+  f << "element vertex " << numVerts << std::endl;
+  f << "property float x" << std::endl;
+  f << "property float y" << std::endl;
+  f << "property float z" << std::endl;
+  f << "property uchar red" << std::endl;
+  f << "property uchar green" << std::endl;
+  f << "property uchar blue" << std::endl;
+  f << "element face " << shapes[0].mesh.num_face_vertices.size() << std::endl;
+  f << "property list uchar int vertex_indices" << std::endl;
+  f << "property ushort object_id" << std::endl;
+  f << "end_header" << std::endl;
+
+  // We need to swap Y and Z to match the .glb files
+  for (size_t i = 0; i < numVerts; i++) {
+    float xyz[3];
+    unsigned char gray[] = {0x80, 0x80, 0x80};
+    xyz[0] = attrib.vertices[i * 3];
+    xyz[1] = -attrib.vertices[i * 3 + 2];
+    xyz[2] = attrib.vertices[i * 3 + 1];
```

Comment:
I am not sure what rotation this is (if it needs to stay, use a named quaternion to do it instead of just hard-coded logic), but we rotate the semantic ply meshes to -Y gravity assuming -Z gravity on load: https://github.com/facebookresearch/habitat-sim/blob/master/src/esp/assets/GenericInstanceMeshData.cpp#L208

Reply:
I've fixed as you suggested. There are a couple of cons to this approach though. It will be slower and you'll lose some precision due to the floating point math. On the other hand, its probably easier to understand.